In [96]:
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans

In [97]:
customers = pd.read_csv('olist_customers_dataset.csv')
orders = pd.read_csv('olist_orders_dataset.csv')
items = pd.read_csv('olist_order_items_dataset.csv')

In [98]:
orders.order_purchase_timestamp = pd.to_datetime(orders.order_purchase_timestamp)
orders.order_delivered_customer_date = pd.to_datetime(orders.order_delivered_customer_date)
orders.order_estimated_delivery_date = pd.to_datetime(orders.order_estimated_delivery_date)
orders.order_approved_at = pd.to_datetime(orders.order_approved_at)
orders.order_delivered_carrier_date = pd.to_datetime(orders.order_delivered_carrier_date)

In [99]:
orders.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0                    2017-10-18  
1                    2018-08-13  
2                    2018-09-04  
3                    2017-12-15  
4                    2018-02-26

In [100]:
customers

customer_id                customer_unique_id  \
0      06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1      18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2      4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3      b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4      4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
...                                 ...                               ...   
99436  17ddf5dd5d51696bb3d7c6291687be6f  1a29b476fee25c95fbafc67c5ac95cf8   
99437  e7b71a9017aa05c9a7fd292d714858e8  d52a67c98be1cf6a5c84435bd38d095d   
99438  5e28dfe12db7fb50a4b2f691faecea5e  e9f50caf99f032f0bf3c55141f019d99   
99439  56b18e2166679b8a959d72dd06da27f9  73c2643a0a458b49f58cea58833b192e   
99440  274fa6071e5e17fe303b9748641082c8  84732c5050c01db9b23e19ba39899398   

       customer_zip_code_prefix          customer_city customer_state  
0                         14409                 franca             SP  
1                          9790  sao bernardo do campo             SP  
2                          1151              sao paulo             SP  
3                          8775        mogi das cruzes             SP  
4                         13056               campinas             SP  
...                         ...                    ...            ...  
99436                      3937              sao paulo             SP  
99437                      6764        taboao da serra             SP  
99438                     60115              fortaleza             CE  
99439                     92120                 canoas             RS  
99440                      6703                  cotia             SP  

[99441 rows x 5 columns]

In [101]:
customers

customer_id                customer_unique_id  \
0      06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1      18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2      4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3      b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4      4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
...                                 ...                               ...   
99436  17ddf5dd5d51696bb3d7c6291687be6f  1a29b476fee25c95fbafc67c5ac95cf8   
99437  e7b71a9017aa05c9a7fd292d714858e8  d52a67c98be1cf6a5c84435bd38d095d   
99438  5e28dfe12db7fb50a4b2f691faecea5e  e9f50caf99f032f0bf3c55141f019d99   
99439  56b18e2166679b8a959d72dd06da27f9  73c2643a0a458b49f58cea58833b192e   
99440  274fa6071e5e17fe303b9748641082c8  84732c5050c01db9b23e19ba39899398   

       customer_zip_code_prefix          customer_city customer_state  
0                         14409                 franca             SP  
1                          9790  sao bernardo do campo             SP  
2                          1151              sao paulo             SP  
3                          8775        mogi das cruzes             SP  
4                         13056               campinas             SP  
...                         ...                    ...            ...  
99436                      3937              sao paulo             SP  
99437                      6764        taboao da serra             SP  
99438                     60115              fortaleza             CE  
99439                     92120                 canoas             RS  
99440                      6703                  cotia             SP  

[99441 rows x 5 columns]

In [102]:
items.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

# 1. Сколько у нас пользователей, которые совершили покупку только один раз? (7 баллов) 

In [103]:
#Покупкой будем считать любой заказ в том случае если пользователь не отменял, 
#в любом другом случае пользователь захотел купить и заказал

In [104]:
orders.merge(customers, on = "customer_id", how = 'outer')\
    .query("order_status != 'canceled'")\
    .groupby("customer_unique_id", as_index = False).agg({"order_id":"count"})\
    .query("order_id == 1")

customer_unique_id  order_id
0      0000366f3b9a7992bf8c76cfdf3221e2         1
1      0000b849f77a49e4a4ce2b2a4ca5be3f         1
2      0000f46a3911fa3c0805444483337064         1
3      0000f6ccb0745a6a4b88665a16c9f078         1
4      0004aac84e0df4da2b147fca70cf8255         1
...                                 ...       ...
95555  fffcf5a5ff07b0908bd4e2dbc735a684         1
95556  fffea47cd6d3cc0a88bd621562a9d061         1
95557  ffff371b4d645b6ecea244b27531430a         1
95558  ffff5962728ec6157033ef9805bacc48         1
95559  ffffd2657e2aad2907e67c3e9daecbeb         1

[92636 rows x 2 columns]

Ответ: 92636

# 2. Сколько заказов в месяц в среднем не доставляется по разным причинам (вывести детализацию по причинам)? (10 баллов)

In [105]:
not_del = orders.query("order_status != 'delivered'")

not_del = not_del.copy()
not_del['year'] = not_del['order_purchase_timestamp'].dt.year
not_del['month'] = not_del['order_purchase_timestamp'].dt.month

In [106]:
not_del

order_id                       customer_id  \
6      136cce7faa42fdb2cefd53fdc79a6098  ed0271e0b7da060a393796590e7b737a   
44     ee64d42b8cf066f35eac1cf57de1aa85  caded193e8e47b8362864762a83db3c5   
103    0760a852e4e9d89eb77bf631eaaf1c84  d2a79636084590b7465af8ab374a8cf5   
128    15bed8e2fec7fdbadb186b57c46c92f2  f3f0e613e0bdb9c7cee75504f0f90679   
154    6942b8da583c2f9957e990d028607019  52006a9383bf149a4fb24226b173106f   
...                                 ...                               ...   
99283  3a3cddda5a7c27851bd96c3313412840  0b0d6095c5555fe083844281f6b093bb   
99313  e9e64a17afa9653aacf2616d94c005b8  b4cd0522e632e481f8eaf766a2646e86   
99347  a89abace0dcc01eeb267a9660b5ac126  2f0524a7b1b3845a1a57fcf3910c4333   
99348  a69ba794cc7deb415c3e15a0a3877e69  726f0894b5becdf952ea537d5266e543   
99415  5fabc81b6322c8443648e1b21a6fef21  32c9df889d41b0ee8309a5efb6855dcb   

      order_status order_purchase_timestamp   order_approved_at  \
6         invoiced      2017-04-11 12:22:08 2017-04-13 13:25:17   
44         shipped      2018-06-04 16:44:48 2018-06-05 04:31:18   
103       invoiced      2018-08-03 17:44:42 2018-08-07 06:15:14   
128     processing      2017-09-03 14:22:03 2017-09-03 14:30:09   
154        shipped      2018-01-10 11:33:07 2018-01-11 02:32:30   
...            ...                      ...                 ...   
99283     canceled      2018-08-31 16:13:44                 NaT   
99313   processing      2018-01-05 23:07:24 2018-01-09 07:18:05   
99347     canceled      2018-09-06 18:45:47                 NaT   
99348  unavailable      2017-08-23 16:28:04 2017-08-28 15:44:47   
99415  unavailable      2017-10-10 10:50:03 2017-10-14 18:35:57   

      order_delivered_carrier_date order_delivered_customer_date  \
6                              NaT                           NaT   
44             2018-06-05 14:32:00                           NaT   
103                            NaT                           NaT   
128                            NaT                           NaT   
154            2018-01-11 19:39:23                           NaT   
...                            ...                           ...   
99283                          NaT                           NaT   
99313                          NaT                           NaT   
99347                          NaT                           NaT   
99348                          NaT                           NaT   
99415                          NaT                           NaT   

      order_estimated_delivery_date  year  month  
6                        2017-05-09  2017      4  
44                       2018-06-28  2018      6  
103                      2018-08-21  2018      8  
128                      2017-10-03  2017      9  
154                      2018-02-07  2018      1  
...                             ...   ...    ...  
99283                    2018-10-01  2018      8  
99313                    2018-02-06  2018      1  
99347                    2018-09-27  2018      9  
99348                    2017-09-15  2017      8  
99415                    2017-10-23  2017     10  

[2963 rows x 10 columns]

In [107]:
not_del = not_del.groupby(["order_status", "year", "month"], as_index = False).agg({"order_id":"count"}).sort_values(['year', 'month'])
not_del

order_status  year  month  order_id
2       canceled  2016      9         2
69       shipped  2016      9         1
3       canceled  2016     10        24
29      invoiced  2016     10        18
50    processing  2016     10         2
..           ...   ...    ...       ...
90       shipped  2018      8        47
112  unavailable  2018      8         7
24      canceled  2018      9        15
91       shipped  2018      9         1
25      canceled  2018     10         4

[113 rows x 4 columns]

In [108]:
not_del.pivot(index=["year", "month"], columns="order_status", values="order_id").fillna(0)

order_status  approved  canceled  created  invoiced  processing  shipped  \
year month                                                                 
2016 9             0.0       2.0      0.0       0.0         0.0      1.0   
     10            0.0      24.0      0.0      18.0         2.0      8.0   
2017 1             0.0       3.0      0.0      12.0         9.0     16.0   
     2             1.0      17.0      0.0      11.0        32.0     21.0   
     3             0.0      33.0      0.0       3.0        23.0     45.0   
     4             1.0      18.0      0.0      14.0        10.0     49.0   
     5             0.0      29.0      0.0      16.0        23.0     55.0   
     6             0.0      16.0      0.0      11.0        12.0     47.0   
     7             0.0      28.0      0.0       7.0        11.0     56.0   
     8             0.0      27.0      0.0      20.0        18.0     41.0   
     9             0.0      20.0      0.0      17.0        22.0     38.0   
     10            0.0      26.0      0.0      16.0        20.0     33.0   
     11            0.0      37.0      2.0      35.0        25.0     72.0   
     12            0.0      11.0      2.0      13.0        35.0     57.0   
2018 1             0.0      34.0      0.0      15.0        29.0     74.0   
     2             0.0      73.0      1.0       6.0         6.0     57.0   
     3             0.0      26.0      0.0      23.0         9.0    133.0   
     4             0.0      15.0      0.0      14.0         8.0     99.0   
     5             0.0      24.0      0.0      24.0         6.0     54.0   
     6             0.0      18.0      0.0       3.0         0.0     43.0   
     7             0.0      41.0      0.0      13.0         1.0     60.0   
     8             0.0      84.0      0.0      23.0         0.0     47.0   
     9             0.0      15.0      0.0       0.0         0.0      1.0   
     10            0.0       4.0      0.0       0.0         0.0      0.0   

order_status  unavailable  
year month                 
2016 9                0.0  
     10               7.0  
2017 1               10.0  
     2               45.0  
     3               32.0  
     4                9.0  
     5               31.0  
     6               24.0  
     7               52.0  
     8               32.0  
     9               38.0  
     10              58.0  
     11              84.0  
     12              42.0  
2018 1               48.0  
     2               30.0  
     3               17.0  
     4                5.0  
     5               16.0  
     6                4.0  
     7               18.0  
     8                7.0  
     9                0.0  
     10               0.0

Вывод:

Мы взяли статус товара как причину того что товар не доставлен. Если товар подтверждет значит счет еще не выставлен, если выставлен счет значит заказ еще в процессе сборки и тд. Исключением явлется отмененный заказа, если отменили то отменили

в сентябре 2016 года 2 заказа отменили, а один заказ еще не доставлен.

в октябре 2016 года 24 заказа отменили, не оплачено 18 заказов, в процессе сборки заказа 2,  отгружен со склада 8, недоступен 7.0

в январе 2017 года 3 заказа отменили, не оплачено 12 заказов, в процессе сборки заказа 9,  отгружен со склада 16, недоступен 10

в феврале 2017 1 заказ только поддтвердили, года 17 заказа отменили, не оплачено 11 заказов, в процессе сборки заказа 32,  отгружен со склада 21, недоступен 45

в марте 2017 года 33 заказа отменили, не оплачено 3 заказов, в процессе сборки заказа 23,  отгружен со склада 45, недоступен 32

в апреле 2017 1 заказ только поддтвердили, года 18 заказа отменили, не оплачено 14 заказов, в процессе сборки заказа 10,  отгружен со склада 49, недоступен 9

в мае 2017 года 29 заказа отменили, не оплачено 16 заказов, в процессе сборки заказа 23,  отгружен со склада 55, недоступен 31

в июне 2017 года 16 заказа отменили, не оплачено 11 заказов, в процессе сборки заказа 12,  отгружен со склада 47, недоступен 24

в июле 2017 года 28 заказа отменили, не оплачено 7 заказов, в процессе сборки заказа 11,  отгружен со склада 56, недоступен 52

в августе 2017 года 27 заказа отменили, не оплачено 20 заказов, в процессе сборки заказа 18,  отгружен со склада 41, недоступен 32

в сентябре 2017 года 20 заказа отменили, не оплачено 17 заказов, в процессе сборки заказа 22,  отгружен со склада 38, недоступен 38

в октябре 2017 года 26 заказа отменили, не оплачено 16 заказов, в процессе сборки заказа 20,  отгружен со склада 33, недоступен 58

в ноябре 2017 года 37 заказа отменили, не оплачено 35 заказов, в процессе сборки заказа 25,  отгружен со склада 72, недоступен 84, 2 заказа было создано клиентом

в декабре 2017 года 11 заказа отменили, не оплачено 13 заказов, в процессе сборки заказа 35,  отгружен со склада 57, недоступен 42, 2 заказа было создано клиентом

в январе 2018 года 34 заказа отменили, не оплачено 15 заказов, в процессе сборки заказа 29,  отгружен со склада 74, недоступен 48

в феврале 2018 года 73 заказа отменили, не оплачено 6 заказов, в процессе сборки заказа 6,  отгружен со склада 57, недоступен 30, 1 заказ было создано клиентом

в марте 2018 года 26 заказа отменили, не оплачено 23 заказов, в процессе сборки заказа 9,  отгружен со склада 133, недоступен 17

в апреле 2018 года 15 заказа отменили, не оплачено 14 заказов, в процессе сборки заказа 8,  отгружен со склада 99, недоступен 5

в мае 2018 года 24 заказа отменили, не оплачено 24 заказов, в процессе сборки заказа 6,  отгружен со склада 54, недоступен 16

в июне 2018 года 18 заказа отменили, не оплачено 3 заказов,  отгружен со склада 43, недоступен 4

в июле 2018 года 41 заказа отменили, не оплачено 13 заказов, в процессе сборки заказа 1,  отгружен со склада 60, недоступен 18

в августе 2018 года 84 заказа отменили, не оплачено 23 заказов,  отгружен со склада 47, недоступен 7.0

в сентябре 2018 года 15 заказа отменили, отгружен со склада 1

в октябре 2018 года 4 заказа отменили

# 3. По каждому товару определить, в какой день недели товар чаще всего покупается. (7 баллов)

In [109]:
item_week = orders.merge(items, on = "order_id").query("order_status != 'canceled'")
item_week['week'] = item_week['order_purchase_timestamp'].dt.day_name()
item_week

order_id                       customer_id  \
0       e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1       53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2       47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3       949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4       ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                  ...                               ...   
112645  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
112646  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
112647  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
112648  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
112649  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

       order_status order_purchase_timestamp   order_approved_at  \
0         delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1         delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2         delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3         delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4         delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   
...             ...                      ...                 ...   
112645    delivered      2018-02-06 12:58:58 2018-02-06 13:10:37   
112646    delivered      2017-08-27 14:46:43 2017-08-27 15:04:16   
112647    delivered      2018-01-08 21:28:27 2018-01-08 21:36:21   
112648    delivered      2018-01-08 21:28:27 2018-01-08 21:36:21   
112649    delivered      2018-03-08 20:57:30 2018-03-09 11:20:28   

       order_delivered_carrier_date order_delivered_customer_date  \
0               2017-10-04 19:55:00           2017-10-10 21:25:13   
1               2018-07-26 14:31:00           2018-08-07 15:27:45   
2               2018-08-08 13:50:00           2018-08-17 18:06:29   
3               2017-11-22 13:39:59           2017-12-02 00:28:42   
4               2018-02-14 19:46:34           2018-02-16 18:17:02   
...                             ...                           ...   
112645          2018-02-07 23:22:42           2018-02-28 17:37:56   
112646          2017-08-28 20:52:26           2017-09-21 11:24:17   
112647          2018-01-12 15:35:03           2018-01-25 23:32:54   
112648          2018-01-12 15:35:03           2018-01-25 23:32:54   
112649          2018-03-09 22:11:59           2018-03-16 13:08:30   

       order_estimated_delivery_date  order_item_id  \
0                         2017-10-18              1   
1                         2018-08-13              1   
2                         2018-09-04              1   
3                         2017-12-15              1   
4                         2018-02-26              1   
...                              ...            ...   
112645                    2018-03-02              1   
112646                    2017-09-27              1   
112647                    2018-02-15              1   
112648                    2018-02-15              2   
112649                    2018-04-03              1   

                              product_id                         seller_id  \
0       87285b34884572647811a353c7ac498a  3504c0cb71d7fa48d967e0e4c94d59d9   
1       595fac2a385ac33a80bd5114aec74eb8  289cdb325fb7e7f891c38608bf9e0962   
2       aa4383b373c6aca5d8797843e5594415  4869f7a5dfa277a7dca6462dcf3b52b2   
3       d0b61bfb1de832b15ba9d266ca96e5b0  66922902710d126a0e7d26b0e3805106   
4       65266b2da20d04dbe00c5c2d3bb7859e  2c9e548be18521d1c43cde1c582c6de8   
...                                  ...                               ...   
112645  f1d4ce8c6dd66c47bbaa8c6781c2a923  1f9ab4708f3056ede07124aad39a2554   
112646  b80910977a37536adeddd63663f916ad  d50d79cb34e38265a8649c383dcffd48   
112647  d1c427060a0f73f6b889a5c7c61f2ac4  a1043bafd471dff536d0c462352beb48   
112648  

In [110]:
week2piv = item_week.groupby(['week', 'product_id'],as_index = False).agg({'order_id':'count'})
week2piv

week                        product_id  order_id
0         Friday  000b8f95fcb9e0096488278317764d19         1
1         Friday  001b72dfd63e9833e8c02742adf472e3         2
2         Friday  002159fe700ed3521f46cfcf6e941c76         1
3         Friday  00250175f79f584c14ab5cecd80553cd         1
4         Friday  002af88741ba70c7b5cf4e4a0ad7ef85         6
...          ...                               ...       ...
60892  Wednesday  ffe013e1b4603e3b0b02fbb159d5b400         1
60893  Wednesday  ffe0fc4e02c3559643ac063fa5cf9d07         1
60894  Wednesday  ffe75578163d45caa52bca3529cb511e         1
60895  Wednesday  fff28f91211774864a1000f918ed00cc         1
60896  Wednesday  fffe9eeff12fcbd74a2f2b007dde0c58         1

[60897 rows x 3 columns]

In [111]:
mostweek = week2piv.sort_values("order_id", ascending = False).drop_duplicates(subset = "product_id")

In [112]:
mostweek[['product_id', 'week', "order_id"]].rename(columns = {'week':"popular_weekday", "order_id": 'orders'})

product_id popular_weekday  orders
53892  422879e10f46682990de24d770e7f83d       Wednesday      93
14460  99a4788cb24856965c36a24e339b6058          Monday      92
39142  aca2eb7d00ea1a7b8ebd4e68314663af        Thursday      89
45142  53b36df67ebb7c41585e8d54d6772e08         Tuesday      76
1889   368c6c730842d78016ad823897a372db          Friday      67
...                                 ...             ...     ...
23854  c73c029653f68c6a83cb371dd148adbd        Saturday       1
23859  c76a7f2aa3596ac11fa33fb6f98cc7c7        Saturday       1
23866  c785fffd281ac9afffc38c8015455fd8        Saturday       1
23868  c7b28fe2ce0fe3c1e58fdb5ccfce9079        Saturday       1
60896  fffe9eeff12fcbd74a2f2b007dde0c58       Wednesday       1

[32735 rows x 3 columns]

# 4. Сколько у каждого из пользователей в среднем покупок в неделю (по месяцам)? Не стоит забывать, что внутри месяца может быть не целое количество недель. Например, в ноябре 2021 года 4,28 недели. И внутри метрики это нужно учесть. (8 баллов) 

In [113]:
ord_date = orders.merge(customers, on = "customer_id", how = 'outer')[['order_id', 'customer_unique_id', 'order_purchase_timestamp']]
ord_date

order_id                customer_unique_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  7c396fd4830fd04220f754e42b4e5bff   
1      53cdb2fc8bc7dce0b6741e2150273451  af07308b275d755c9edb36a90c618231   
2      47770eb9100c2d0c44946d9cf07ec65d  3a653a41f6f9fc3d2a113cf8398680e8   
3      949d5b44dbf5de918fe9c16f97b45f8a  7c142cf63193a1473d2e66489a9ae977   
4      ad21c59c0840e6cb83a9ceb5573f8159  72632f0f9dd73dfee390c9b22eb56dd6   
...                                 ...                               ...   
99436  9c5dedf39a927c1b2549525ed64a053c  6359f309b166b0196dbf7ad2ac62bb5a   
99437  63943bddc261676b46f01ca7ac2f7bd8  da62f9e57a76d978d02ab5362c509660   
99438  83c1379a015df1e13d02aae0204711ab  737520a9aad80b3fbbdad19b66b37b30   
99439  11c177c8e97725db2631073c19f07b62  5097a5312c8b157bb7be58ae360ef43c   
99440  66dea50a8b16d9b4dee7af250b4be1a5  60350aa974b26ff12caad89e55993bd6   

      order_purchase_timestamp  
0          2017-10-02 10:56:33  
1          2018-07-24 20:41:37  
2          2018-08-08 08:38:49  
3          2017-11-18 19:28:06  
4          2018-02-13 21:18:39  
...                        ...  
99436      2017-03-09 09:54:05  
99437      2018-02-06 12:58:58  
99438      2017-08-27 14:46:43  
99439      2018-01-08 21:28:27  
99440      2018-03-08 20:57:30  

[99441 rows x 3 columns]

In [114]:
ord_date['week'] = ord_date['order_purchase_timestamp'].dt.week

C:\Users\nurbe\AppData\Local\Temp\ipykernel_4316\1119107358.py:1: FutureWarning: Series.dt.weekofyear and Series.dt.week have been deprecated. Please use Series.dt.isocalendar().week instead.
  ord_date['week'] = ord_date['order_purchase_timestamp'].dt.week


In [115]:
ord_date.groupby(['customer_unique_id', 'week'], as_index = False)\
    .agg({'order_id':'count'})\
    .sort_values("order_id", ascending = False)

customer_unique_id  week  order_id
7341   12f5d6e1cbf93dafd9dcc19095df0b3d     1         6
14527  25a560b9a6006157838aab1bdbd68624    14         4
50771  83e7958a94bd7f74a9414d8782f87628     1         4
69611  b4e4f24de1e8725b74e4a1f4975116ed     8         4
77026  c8460e4251689ba205045f3ea17884a1    32         4
...                                 ...   ...       ...
32927  55845d91497951227fd70e450ffb35e0     6         1
32926  55840dd9a1ca54d819ce02d06bacde82    20         1
32925  55836600036b58714e3e52f3921fc36d    19         1
32924  558308f294848bcacb6097d28a2837d4    26         1
98348  ffffd2657e2aad2907e67c3e9daecbeb    18         1

[98349 rows x 3 columns]

# 5. Используя pandas, проведи когортный анализ пользователей. В период с января по декабрь выяви когорту с самым высоким retention на 3й месяц.(15 баллов)

In [116]:
import pandas as pd

customers = pd.read_csv('olist_customers_dataset.csv')
orders = pd.read_csv('olist_orders_dataset.csv')
items = pd.read_csv('olist_order_items_dataset.csv')

orders.order_purchase_timestamp = pd.to_datetime(orders.order_purchase_timestamp)
orders.order_delivered_customer_date = pd.to_datetime(orders.order_delivered_customer_date)
orders.order_estimated_delivery_date = pd.to_datetime(orders.order_estimated_delivery_date)
orders.order_approved_at = pd.to_datetime(orders.order_approved_at)
orders.order_delivered_carrier_date = pd.to_datetime(orders.order_delivered_carrier_date)

In [117]:
# Объединение таблиц
merged_data = pd.merge(items, orders, on="order_id")
merged_data = pd.merge(merged_data, customers, on="customer_id")


In [118]:
# преобразование дат в нужный формат
merged_data['order_purchase_timestamp'] = pd.to_datetime(merged_data['order_purchase_timestamp'])


In [119]:
# Извлечение месяца первого заказа для каждого клиента
merged_data['order_month'] = merged_data['order_purchase_timestamp'].dt.to_period('M')
merged_data['cohort_month'] = merged_data.groupby('customer_unique_id')['order_month'].transform('min')

In [120]:
# время жизни когорт в месяцах
merged_data['cohort_lifetime'] = (merged_data['order_month'].dt.year - merged_data['cohort_month'].dt.year) * 12 + (merged_data['order_month'].dt.month - merged_data['cohort_month'].dt.month) + 1


In [121]:
# сводная таблица с количеством уникальных клиентов в каждой когорте и месяце
cohort_counts = merged_data.groupby(['cohort_month', 'cohort_lifetime']).agg({'customer_unique_id': 'nunique'}).reset_index()

In [122]:
# общее количество клиентов в первом месяце каждой когорты
initial_customer_count = cohort_counts[cohort_counts['cohort_lifetime'] == 1][['cohort_month', 'customer_unique_id']]

In [123]:
initial_customer_count = initial_customer_count.rename(columns={'customer_unique_id': 'cohort_customer_count'})

In [124]:
cohort_counts = pd.merge(cohort_counts, initial_customer_count, on='cohort_month')

In [125]:
# retention rate
cohort_counts['retention'] = cohort_counts['customer_unique_id'] / cohort_counts['cohort_customer_count']


In [126]:
# Выберите когорту с самым высоким retention на 3-й месяц
highest_retention_cohort = cohort_counts[cohort_counts['cohort_lifetime'] == 3].sort_values(by='retention', ascending=False).iloc[0]

highest_retention_cohort


cohort_month              2017-09
cohort_lifetime                 3
customer_unique_id             22
cohort_customer_count        4090
retention                0.005379
Name: 143, dtype: object

# 6. Часто для качественного анализа аудитории использую подходы, основанные на сегментации. Используя python, построй RFM-сегментацию пользователей, чтобы качественно оценить свою аудиторию. В кластеризации можешь выбрать следующие метрики: R - время от последней покупки пользователя до текущей даты, F - суммарное количество покупок у пользователя за всё время, M - сумма покупок за всё время. Подробно опиши, как ты создавал кластеры. Для каждого RFM-сегмента построй границы метрик recency, frequency и monetary для интерпретации этих кластеров. Пример такого описания: RFM-сегмент 132 (recency=1, frequency=3, monetary=2) имеет границы метрик recency от 130 до 500 дней, frequency от 2 до 5 заказов в неделю, monetary от 1780 до 3560 рублей в неделю. Описание подхода можно найти тут. (23 балла)

In [127]:
import pandas as pd

# Загрузка данных
customers = pd.read_csv('olist_customers_dataset.csv')
orders = pd.read_csv('olist_orders_dataset.csv')
items = pd.read_csv('olist_order_items_dataset.csv')

orders.order_purchase_timestamp = pd.to_datetime(orders.order_purchase_timestamp)
orders.order_delivered_customer_date = pd.to_datetime(orders.order_delivered_customer_date)
orders.order_estimated_delivery_date = pd.to_datetime(orders.order_estimated_delivery_date)
orders.order_approved_at = pd.to_datetime(orders.order_approved_at)
orders.order_delivered_carrier_date = pd.to_datetime(orders.order_delivered_carrier_date)


In [128]:
# Объединение данных
merged_data = pd.merge(customers, orders, on="customer_id")
merged_data = pd.merge(merged_data, items, on="order_id")

In [129]:
merged_data["order_purchase_timestamp"] = pd.to_datetime(merged_data["order_purchase_timestamp"])
merged_data["order_delivered_customer_date"] = pd.to_datetime(merged_data["order_delivered_customer_date"])

In [130]:
# Рассчет Recency, Frequency и Monetary
current_date = pd.to_datetime("2023-11-13")
rfm_table = merged_data.groupby("customer_id").agg({
    'order_purchase_timestamp': lambda x: (current_date - x.max()).days,
    'order_id': 'count',
    'price': 'sum'
})

rfm_table.rename(columns={'order_purchase_timestamp': 'recency',
                          'order_id': 'frequency',
                          'price': 'monetary'}, inplace=True)

In [131]:
scaler = StandardScaler()
rfm_scaled = scaler.fit_transform(rfm_table)
rfm_scaled = pd.DataFrame(rfm_scaled, index=rfm_table.index, columns=rfm_table.columns)


In [132]:
# Определение числа кластеров (может потребоваться подобрать оптимальное значение)
kmeans = KMeans(n_clusters=3, random_state=42)
rfm_table['cluster'] = kmeans.fit_predict(rfm_scaled)


In [133]:
cluster_summary = rfm_table.groupby('cluster').agg({
    'recency': ['min', 'max'],
    'frequency': ['min', 'max'],
    'monetary': ['min', 'max']
}).reset_index()

# Вывод интерпретации
for i, row in cluster_summary.iterrows():
    print(f"Cluster {int(row['cluster'])}:")
    print(f"Recency: {int(row['recency']['min'])} to {int(row['recency']['max'])} days")
    print(f"Frequency: {int(row['frequency']['min'])} to {int(row['frequency']['max'])} orders")
    print(f"Monetary: {int(row['monetary']['min'])} to {int(row['monetary']['max'])} monetary units")
    print("\n")


Cluster 0:
Recency: 1896 to 2161 days
Frequency: 1 to 2 orders
Monetary: 0 to 864 monetary units


Cluster 1:
Recency: 2161 to 2625 days
Frequency: 1 to 2 orders
Monetary: 2 to 949 monetary units


Cluster 2:
Recency: 1901 to 2614 days
Frequency: 1 to 21 orders
Monetary: 9 to 13440 monetary units


